### DISCUSSION TITLE
Reloading after deploy, "high" availability

### DISCUSSION THREAD

**Discussion by aleho** [php/frankenphp#1025] [General]
Lately I've been thinking about a few ways to reload code / clear the OPCache after deploys.

As FrankenPHP and Caddy are tightly coupled, reloading FrankenPHP will drop all active connections to clients.

Looking at Deployer, they [used to recommend](https://deployer.org/docs/7.x/avoid-php-fpm-reloading) setting `opcache.validate_timestamps = 1`, but have recently implemented [cachetool](https://deployer.org/docs/7.x/contrib/cachetool) to clear the OPCache in FPM through a socket after a deploy. Of course this is not available for FrankenPHP.

What are your ideas to achieve similar "zero-downtime" deploys for FrankenPHP in worker mode?

I could still add a "reload-frankenphp" script to my containers and live with the short (or sometimes longer) downtime, but is there a better way? Maybe related to https://github.com/dunglas/frankenphp/discussions/556#discussioncomment-8386615?

**Comment by AlliBalliBaba** [comment]
The [file watcher] (https://github.com/dunglas/frankenphp/pull/1013) is being worked on, but it will still take some time to complete.
I think our interests here align a bit since I'm also trying to make reloading workers more graceful and am looking into clearing opcache along the way.

For a deployment it probably would make more sense to expose the same reload behavior some other way (maybe via caddy's admin?)

**Reply by aleho** [reply]
I actually really don't mind any "manual" work on deploy, I wouldn't even need the file watcher (though it'd be absolutely great if that was a feature of FrankenPHP)!

So, if there's no inotify for whatever reason – I don't think it would have an impact on my projects, but YMMV – I'd be completely fine with calling a FrankenPHP admin endpoint inside the container to "purge" OPCache via `opcache_reset()`.

**Reply by aleho** [reply]
Now that watchers are released and work very well for dev environments, do you have any suggestions for live deployments? (ping @Alliballibaba2)

**Comment by AlliBalliBaba** [comment]
#1266 will add an admin api endpoint for graceful reloading that will also clear opcache (probably better than using the watcher in prod)
`curl -X POST http://localhost:2019/frankenphp/workers/reload`.

**Reply by aleho** [reply]
Ah! Great! I subscribed to this PR but I missed that it also has an endpoint that is not for scaling.

Thank you! This will be a great addition for live deploys.
